appshell/appshell_extensions_mac.mm

```diff
@@ -644,52 +647,110 @@ int32 GetFileInfo(ExtensionString filename, uint32& modtime, bool& isDir, double
+            try {
+                if (encoding == "UTF-8") {
+                    GetCharsetMatch(contents.c_str(), contents.size(), detectedCharSet);
+                }
+                else {
+                    detectedCharSet = encoding;
+                }
+                if (!detectedCharSet.empty()) {
+                    std::transform(detectedCharSet.begin(), detectedCharSet.end(), detectedCharSet.begin(), ::toupper);
+                    UnicodeString ustr(contents.c_str(), detectedCharSet.c_str());
+                    UErrorCode status = U_ZERO_ERROR;
+                    int targetLen = ustr.extract(NULL, 0, NULL, status);
+                    if(status != U_BUFFER_OVERFLOW_ERROR) {
+                        return ERR_UNSUPPORTED_ENCODING;
+                    }
+                    std::auto_ptr<char> target(new char[targetLen + 1]());
+                    status = U_ZERO_ERROR;
+                    ustr.extract(target.get(), targetLen, NULL, status);
+                    target.get()[targetLen] = '\0';
+                    if (U_SUCCESS(status)) {
+                        contents.assign(target.get(), targetLen);
+                        encoding = detectedCharSet;
+                        return NO_ERROR;
+                    }
+                    else {
+                        return ERR_UNSUPPORTED_ENCODING;
+                    }
+                }
+                else {
+                    return ERR_UNSUPPORTED_ENCODING;
+                }
+            } catch (...) {
+                return ERR_UNSUPPORTED_ENCODING;
+            }
+        } catch (...) {
+            return ERR_CANT_READ;
+        }
     }
     
     return ConvertNSErrorCode(error, true);
 }
 
 int32 WriteFile(ExtensionString filename, std::string contents, ExtensionString encoding)
 {
-    NSString* path = [NSString stringWithUTF8String:filename.c_str()];
-    NSString* contentsStr = [NSString stringWithUTF8String:contents.c_str()];
-    NSStringEncoding enc;
+    const char *filenameStr = filename.c_str();
     NSError* error = nil;
+    if (encoding == "utf8") {
+        encoding = "UTF-8";
+    }
     
-    if (encoding == "utf8")
-        enc = NSUTF8StringEncoding;
-    else
-        return ERR_UNSUPPORTED_ENCODING;
-    
-    const NSData* encodedContents = [contentsStr dataUsingEncoding:enc];
-    NSUInteger len = [encodedContents length];
-    NSOutputStream* oStream = [NSOutputStream outputStreamToFileAtPath:path append:NO];
-    
-    [oStream open];
-    NSInteger res = [oStream write:(const uint8_t*)[encodedContents bytes] maxLength:len];
-    [oStream close];
+    if (encoding != "UTF-8") {
+        UErrorCode status = U_ZERO_ERROR;
+        UConverter *conv = ucnv_open(encoding.c_str(), &status);
+        if (U_FAILURE(status)) {
+            return ERR_CANT_WRITE;
+        }
+        UnicodeString ustr(contents.c_str());
+        int targetLen = ustr.extract(NULL, 0, conv, status);
+        if(status != U_BUFFER_OVERFLOW_ERROR) {
```

Comment:
ucnv_close() is not called if you return from here. implement proper stack based cleanups